File: 24.cpp
#include <vector>
#include <deque>
#include <string>
#include <variant>
#include <iostream>
#include <algorithm>
#include <unordered_map>
#include <fstream>
#include <sstream>

using namespace std;

int mem[4];

int charToIndex(char c) {
    if (c == 'x') {
        return 0;
    } else if (c == 'y') {
        return 1;
    } else if (c == 'z') {
        return 2;
    } else if (c == 'w') {
        return 3;
    }
    return -1;
}

void add(char a, int b) {
    mem[charToIndex(a)] += b;
}

void mul(char a, int b) {
    mem[charToIndex(a)] *= b;
}

void div(char a, int b) {
    mem[charToIndex(a)] /= b;
}

void mod(char a, int b) {
    mem[charToIndex(a)] %= b;
}

void eql(char a, int b) {
    mem[charToIndex(a)] = mem[charToIndex(a)] == b;
}

struct Command {
    string command;
    char reg;
    variant<char, int> val;
};

vector<Command> commands;

vector<string> split(const string& str, char delimiter) {
    vector<string> tokens;

    stringstream stream(str);
    string token;
    while (getline(stream, token, delimiter)) {
        tokens.push_back(token);
    }

    return tokens;
}

void parseFile() {
    ifstream file("24.txt");
    string line;
    if (file.is_open()) {
        while (getline(file, line)) {
            vector<string> tokens = split(line, ' ');
            if (tokens[0] == "inp") {
                commands.push_back(Command{"inp", tokens[1][0]});
            } else {
                variant<char, int> val;
                if (charToIndex(tokens[2][0]) == -1) {
                    val = stoi(tokens[2]);
                } else {
                    val = tokens[2][0];
                }
                commands.push_back(Command{tokens[0], tokens[1][0], val});
            }
        }
    } else {
        cout << "File not found >:(" << endl;
    }
}

void exec(const vector<Command>& commands, deque<int> inputs, int z) {
    for (int i = 0; i < 4; i++) mem[i] = 0;
    mem[charToIndex('z')] = z;

    for (const Command& command : commands) {
        if (command.command == "inp") {
            mem[charToIndex(command.reg)] = inputs[0];
            inputs.pop_front();
        } else {
            int b = holds_alternative<int>(command.val) ? get<int>(command.val) : mem[charToIndex(get<char>(command.val))];
            if (command.command == "add") {
                add(command.reg, b);
            } else if (command.command == "mul") {
                mul(command.reg, b);
            } else if (command.command == "div") {
                div(command.reg, b);
            } else if (command.command == "mod") {
                mod(command.reg, b);
            } else if (command.command == "eql") {
                eql(command.reg, b);
            }
        }
    }
}

vector<Command> getProgramForDigitIndex(int digitIndex) {
    vector<Command> program;

    int count = 0;
    for (int i = (int)commands.size() - 1; i >= 0; i--) {
        const Command& c = commands[i];
        program.push_back(c);
        if (c.command == "inp") {
            if (++count == 14 - digitIndex) {
                break;
            } else {
                program.clear();
            }
        }
    }

    reverse(program.begin(), program.end());
    return program;
}

vector<Command> digitPrograms[14];
void makeDigitPrograms() {
    for (int i = 0; i < 14; i++) digitPrograms[i] = getProgramForDigitIndex(i);
}

struct hash_pair {
    template <class T1, class T2>
    size_t operator()(const pair<T1, T2>& p) const
    {
        auto hash1 = hash<T1>{}(p.first);
        auto hash2 = hash<T2>{}(p.second);
        return hash1 ^ hash2;
    }
};

unordered_map<pair<int, int>, pair<string, bool>, hash_pair> dp;
pair<string, bool>& solve(int digitIndex, int z, bool biggest) {
    pair<int, int> state = make_pair(digitIndex, z);
    if (dp.find(state) != dp.end()) return dp[state];

    int start = biggest ? 9 : 1;
    int dir = biggest ? -1 : 1;
    for (int i = start; i >= 1 && i <= 9; i += dir) {
        exec(digitPrograms[digitIndex], {i}, z);
        int startZ = mem[charToIndex('z')];
        string allDigits;
        bool endsWithZero;
        if (digitIndex == 13) {
            allDigits = "";
            endsWithZero = startZ == 0;
        } else {
            pair<string, bool>& result = solve(digitIndex + 1, startZ, biggest);
            allDigits = result.first;
            endsWithZero = result.second;
        }
        if (endsWithZero) {
            dp[state] = make_pair(to_string(i) + allDigits, true);
            return dp[state];
        }
    }

    if (dp.find(state) == dp.end()) dp[state] = make_pair("", false);
    return dp[state];
}

int main() {
    parseFile();
    makeDigitPrograms();
    cout << "Part 1: " << solve(0, 0, true).first << endl;
    dp.erase(make_pair(0, 0));
    cout << "Part 2: " << solve(0, 0, false).first << endl;
    return 0;
}

File: 24.js
pipe=(start=document.body.innerText)=>(...fns)=>fns.reduce((input, fn) => fn(input), start)
function permutator(inputArr) {
    var results = [];
  
    function permute(arr, memo) {
      var cur, memo = memo || [];
  
      for (var i = 0; i < arr.length; i++) {
        cur = arr.splice(i, 1);
        if (arr.length === 0) {
          results.push(memo.concat(cur));
        }
        permute(arr.slice(), memo.concat(cur));
        arr.splice(i, 0, cur[0]);
      }
  
      return results;
    }
  
    return permute(inputArr);
  }
getNums = txt => txt.match(/(\d+)/g).map(Number)
makeFreqArr = nums => nums.reduce((r, c) => ++r[c]&&r, Array(9).fill(0))
repeat = (times, ...fns) => (input) => { for (let i = 0; i < times; i++) input = fns.reduce((input, fn) => fn(input)||input, input); return input }
rotateLeft = (nums) => void(nums.push(nums.shift()))
rotateRight = (nums) => void(nums.unshift(nums.pop()))
sum=a=>a.reduce((x,y)=>x+y,0)
arrwithdefault=(defaultval=()=>0)=>new Proxy([], {
  get(a,i,r) {
    for (let j = a.length; j < i; j++) r[j]; // set prev defaults
    if (a[i] === undefined) {
      a[i] = defaultval()
    }
    return a[i]
  },
  set(a,i,v,r) {
    for (let j = a.length; j < i; j++) r[j]; // set prev defaults
    a[i] =v
  }
})
imat=(dims,defaultval=0)=>{
  res=[()=>defaultval]
  for (let i = 0; i < dims; i++) {
    res.push(()=>arrwithdefault(res[i]))
  }
  return res[res.length-1]()
}
// mat = (dims,fill=()=>0) => dims.length === 1 ? Array(dims[0]).fill(fill()):Array.from({length:dims.splice(-1)}, () => mat(dims,fill))
zip = (a,...r)=> a.map((a,i)=>[a,...r.map(r=>r[i])])
rng = (f, t) => Array.from({length:Math.abs(f-t)+1},(_,i)=>f+i*(f<t?1:-1))
repeat=(arr,cnt)=>Array.from({length:cnt},()=>[...arr]).flat(1)
forcelen=(arr,len)=>repeat(arr,Math.ceil(len/arr.length)).slice(0, len)

f = {
  numberlist: /(\d+)/g,
  number: /(\d+)$/g,
  digits: /(\d)/g,
  chars: /(\w)/g,
  things: /(.)/g,
  rlines: '',//document.body.innerText.trim().split('\n'),
  lines: ()=>f.rlines.map(e=>e.trim()).filter(e=>e),
  read: (r=f.numberlist,lines=f.lines())=>lines.map(l=>[...l.matchAll(r)].map(e=>e.slice(1).map(e=>isNaN(e)?e:Number(e))).flat(Infinity)),
  // reads matrix based on given size (height)
  readmatsCS: (size,lines=f.lines()) => f.read(f.numberlist,lines).reduce((mats, row, i, a) => i%size==0?[...mats,Array.from({length:size},(_,j)=>a[i + j])]:mats, []),
  // reads matrix based on empty lines
  readmatsEL: (lines=f.rlines) => lines.reduce((r, line) => line.trim().length ? [r, r[1].push([...line.matchAll(f.numberlist)].map(e=>e.slice(1).map(e=>isNaN(e)?e:Number(e))).flat(Infinity))][0] : [r, r[0].push(r[1]), r[1] = []][0], [[], []])[0]
}
Set.prototype.union = function (...sets) {
  return new Set([...this, ...sets.flatMap(set => [...set])])
}
Set.prototype.intersect = function (...sets) {
  return new Set([...this].filter(e => sets.every(set => set.has(e))))
}
Set.prototype.difference = function (b) {
  return new Set([...this].filter(e => !b.has(e)))
}
Set.prototype.xor = function (b) {
  return this.union(this.difference(b), b.difference(this))
}
set = (...elements) => new Set(elements)

b = require('fs').readFileSync('./24.txt', { encoding: 'utf8' })
// b=document.body.innerText.trim()
const lines = b.trim().split('\n').map(e => e.match(/\w+|-?\d+/g).map((e,i) => isNaN(e) ? e : Number(e)))

let env = {}
// a is always var. b is var or number
const action = {
  add(a, b) {
    env[a] += b
  },
  mul(a, b) {
    env[a] *= b
  },
  inp(name, val) {
    env[name] = val
  },
  div(a, b) {
    if (b === 0) throw new Error('shouldnt happen div')
    env[a] /= b
    env[a] = Math.floor(env[a])
  },
  mod(a, b) {
    if (a < 0 || b < 0) throw new Error('shouldnt happen mod')
    env[a] = env[a] % b
  },
  eql(a, b) {
    env[a] = Number(env[a] === b)
  }
}
function exec(lines, inputs, z = 0) {
  env.x = 0
  env.y = 0
  env.z = z
  env.w = 0

  inputs = [...inputs].map(Number)
  for (let [command, a, b] of lines) {
    if (b) {
      if (isNaN(b)) b = env[b]
      else b = parseInt(b)
    }
    
    if (command === 'inp') {
      console.log('z',env.z)
      if (inputs.length === 0) throw new Error('not enough inputs')
      action.inp(a, inputs.shift())
    } else action[command](a, b)
  }

  return env
}

// get program for a specific digit (digit index is left to right, 0 indexed)
function getProgramForDigitIndex(digitIndex) {
  let newLines = []

  let count = 0
  for (let i = lines.length - 1; i >= 0; i--) {
    newLines.push(lines[i])
    if (lines[i][0] === 'inp') {
      if (++count === 14 - digitIndex) {
        break
      } else {
        newLines = []
      }
    }
  }
  
  return newLines.reverse()
}

// maps [digitIndex, startZ] to => [allDigits, endsWith0]
const mem = {}

function solve(digitIndex, z = 0) {
  if (digitIndex > 13) return ['', z === 0]
  if (mem[[digitIndex, z]]) return mem[[digitIndex, z]]

  // for (let i = 9; i >= 1; i--) {
  for (let i = 1; i <= 9; i++) {
    const { z: startZ } = exec(getProgramForDigitIndex(digitIndex), [i], z)
    const [allDigits, endsWithZero] = solve(digitIndex + 1, startZ)
    if (endsWithZero) {
      mem[[digitIndex, z]] = [i.toString() + allDigits, true]
      return mem[[digitIndex, z]]
    }
  }
  
  mem[[digitIndex, z]] ??= ['', false]
  return mem[[digitIndex, z]]
}


const start = 0 // at which index to start (just to make sure it works on small examples)
// const [p1, z] = solve(start)
// console.log(p1, z)
console.log('z', exec(rng(start, 13).flatMap(i => getProgramForDigitIndex(i)), '59996912981939', 0))


// highest: 59996912981939 (no time, too long in js)
// lowest: 17241911811915 (1m08s)

// total time in cpp: 3m54s